SOC-158 add build.sbt

henrytill opened the pull request:
**6/17: THIS IS NOW READY FOR REVIEW.**

https://synereo.atlassian.net/browse/SOC-158

Here's my initial work porting the build to sbt.  It currently uses the same compiler options as the Maven build, and should produce the exact same set of warnings (I have diffed their outputs to verify this). 

I've built the `prolog4j` JARs from our fork and vendored them as unmanaged dependencies for now.  See my included [notes](https://github.com/henrytill/specialk/blob/sbtize/doc/notes.md) about them and the other unmanaged dependencies.  We could consider hosting these through [bintray](https://bintray.com/), though whether we ultimately choose to fetch or vendor them is an open question from a security standpoint.

Before merging, I'd like to finish the following tasks:
- [x] Get `sbt-assembly` working so we can create a fat/uber JAR.
- [x] Bring back [JLex](https://www.cs.princeton.edu/~appel/modern/java/JLex/) (perhaps just by vendoring the source, considering it's a single file).
- [x] Port the BNFC code generation tasks to sbt.

cc @leithaus @n10n @jborozdina @stephensong 

Questions/comments/suggestions welcome.

review-ninja commented:
<a href="https://app.review.ninja/synereo/specialk/pull/5" target="_blank"><img src="https://app.review.ninja/29917615/pull/5/badge" alt="ReviewNinja"/>

henrytill commented:
I've built the following with JDK 7u95:
```
lib/prolog4j-api-0.2.1-SNAPSHOT.jar
lib/prolog4j-tuprolog-0.2.1-SNAPSHOT.jar
```

henrytill commented:
I've added the following four sbt tasks for BNFC code generation:

- `bnfcGenerate`
- `bnfcClean`
- `bnfcGenerateDocs`
- `bnfcCleanDocs`

These use the following settings:
- `bnfcNamespace`, which defaults to `"com.biosimilarity.lift.model"`
- `bnfcGrammarDir`, which defaults to `./src/main/bnfc`
- `bnfcOutputDir`, which defaults to `./src/main/java`
- `bnfcDocDir`, which defaults to `./doc/bnfc`

These tasks replace the old Ant `build.xml` for BNFC code generation, and allow us to avoid generating and using a `Makefile` with BNFC.  I've left out the Haskell code generation and LaTeX to HTML conversion tasks.  They can be added later if necessary.

In order to use them, a `bnfc` executable must be available on `$PATH`.

**NOTE:** For now, I've decided to leave the original files generated by BNFC (and processed by JLex and CUP) intact.  I've introduced these tasks for use in later refactorings, so that we can disentangle the generated code from the rest of the codebase.
